### ISSUE TITLE
test_install.py failing and virtualenv issues

### ISSUE THREAD

**Issue by abravalheri** [pyscaffold/pyscaffold#238]
(Related to  #119)

The following command is currently failing:

```bash
$ tox -e default -- tests/test_install.py

# ...

tests/test_install.py::test_sdist_install FAILED                         [  6%]
tests/test_install.py::test_sdist_install_dirty FAILED                   [ 12%]
tests/test_install.py::test_sdist_install_with_1_0_tag FAILED            [ 18%]
tests/test_install.py::test_sdist_install_with_1_0_tag_dirty FAILED      [ 25%]
tests/test_install.py::test_bdist_install FAILED                         [ 31%]
tests/test_install.py::test_bdist_wheel_install FAILED                   [ 37%]
tests/test_install.py::test_git_repo FAILED                              [ 43%]
tests/test_install.py::test_git_repo_dirty FAILED                        [ 50%]
tests/test_install.py::test_git_repo_with_1_0_tag FAILED                 [ 56%]
tests/test_install.py::test_git_repo_with_1_0_tag_dirty FAILED           [ 62%]
tests/test_install.py::test_sdist_install_with_data FAILED               [ 68%]
tests/test_install.py::test_bdist_install_with_data FAILED               [ 75%]
tests/test_install.py::test_bdist_wheel_install_with_data FAILED         [ 81%]
tests/test_install.py::test_edit_install_with_data FAILED                [ 87%]
tests/test_install.py::test_setup_py_install_with_data FAILED            [ 93%]
tests/test_install.py::test_setup_py_develop_with_data FAILED            [100%]
```

### Steps to reproduce:
1. Clone PyScaffold v4.0.x branch
2. Comment the extra dependencies for `all` in `setup.cfg` (required since extensions are not updated yet.
3. Run tox

### Versions and main components
* PyScaffold Version: v4.0.x head of the branch
* Python Version: 3.6.8
* Operating system: Ubuntu
* How did you install PyScaffold: pipx

### Meaningful pieces of trace:
```
_______________________ test_sdist_install_with_1_0_tag ________________________

demoapp = <tests.test_install.DemoApp object at 0x7f26e0eaa550>

    def test_sdist_install_with_1_0_tag(demoapp):
        (demoapp
            .make_dirty_tree()
            .make_commit()
            .tag('v1.0', 'final release')
>           .build('sdist')
            .install())

$HOME/pyscaffold/tests/test_install.py:234: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
$HOME/pyscaffold/tests/test_install.py:139: in build
    self.run('python', 'setup.py', dist)
$HOME/pyscaffold/tests/test_install.py:123: in run
    return self.venv.run(args, **kwargs).strip()
$HOME/pyscaffold/.tox/default/lib/python3.6/site-packages/pytest_virtualenv.py:151: in run
    return super(VirtualEnv, self).run(args, **kwargs)
....
  File "/tmp/easy_install-rTQsTF/PyScaffold-3.1/src/pyscaffold/utils.py", line 307
    major, minor, *rest = (parse_version(pyscaffold_version)
                  ^
SyntaxError: invalid syntax
```

### Diagnostics
It seems that `pytest-virtualenv` is using the system-wide `python` executable (which happens to be 2.7 in my computer) as basis for creating the virtualenv, instead of the currently running version.

I have made an attempt to solve this issue, by relying on stdlib's `venv` instead of `pytest-virtualenv`, on a branch [`fix-venv-tests`](https://github.com/abravalheri/pyscaffold/blob/fix-venv-tests/tests/system/__init__.py) in my personal repo. (I vaguely remember we also had some issues related to `virtualenv`...)
However, I came across another issue: it seems that `venv`, `pip`, `ensurepip` and even [`get-pip.py`](https://bootstrap.pypa.io/get-pip.py) do not install pip inside a venv generated using a python executable from another venv (the one in `.tox` in our case).
This seems to be related to [an issue discussed in the pip repository](https://github.com/pypa/pip/issues/6355), but I think all boils down to `ensurepip` (that seems to ignore `--root` and even the `-I` flag in python).

I suspect that even if we manage to solve the issue using `pytest-virtualenv` it might be the case the test isolation would be compromised (if they are relying in ensure pip also).

**Comment by FlorianWilhelm** [comment]
Thanks @abravalheri for finding this out and providing a PR so fast. Just to make sure it get this right, is it actually a bug in `pip`? Since your PR adds much code and basically reproduces parts of `pytest-virtualenv` I wonder if fixing the bug at the source would be better since the complexity of the code increases a lot. What do you think? We should at least make sure this can be easily rolled back and is documented when the actual root cause is fixed.

**Comment by abravalheri** [comment]
Hi @FlorianWilhelm, thanks for your comments!

After a while playing around this issue, I have the following guesses:
1. By default, `pytest-virtualenv` uses the `python` binary without appending a version string to it
    - It seems that this can be solved by tapping into the internals of `pytest-virtualenv`: instead of using the provided fixture, we use. However there are no documentation about this in the library, so I would consider it as invading an private API (although the class is not marked as private explicitly)
2. Python flag `-I` seems to ignore virtual envs, as long as they are created with stdlib's `venv`, not the ones created with `virtualenv` (which is the case when using `tox`).

Please have in mind that (1) is related to keep using `pytest-virtualenv` and (2) is related to switching over `venv`.

So, I can see two ways forward: by using `pytest-virtualenv` as described in (1) or by using something like #239 is provided.

One of the pros of using `pytest-virtualenv` is that we offload the maintenance to the python package as you have mentioned. However this is also part of the cons: we will be tapping into the internals of the package and then if we have issues, it is much harder to solve (as for example the one that we are having right now) and then we have to dig into `pytest-virtualenv` and its dependencies (e.g. `pytest-shutil`)

The pros of #239 is that, if we have issues, we have to check one single file so ~200 lines long, and we have the complete control of it. The main con is that we lose the ability to run commands with coverage.

Regarding the complexity of #239, it is always good do avoid adding complexity, but to be fair, the majority of the code in `tests/system/__init__.py` is coming from other places (re-organization). The big additions are `global_python` and parts of new class `Venv`, since things like `venv_run` already existed before. It is a bit of trade off, increases complexity on that file, but remvoes from other places.

Please notice that I am still iterating and simplifying the implementation after isolating the issue using [this](https://virtualenv.pypa.io/en/stable/reference/#compatibility-with-the-stdlib-venv-module) as a reference. If you prefer following the approach in (1) please let me know so I switch directions.

**Comment by FlorianWilhelm** [comment]
Hi @abravalheri, sorry for the late reply. I ran and reproduced this bug, also took a look into `pytest-virtualenv`. Let's recap and make sure that we are on the same page here. Please correct me if any of the following statements are wrong:
1. Tox creates a virtualenv with a specified Python version, e.g. Python 3.6 or Python 3.7
2. Within such a virtualenv we run our unit tests with py.test
3. If a unit tests creates its own virtualenv by using `pytest-virtualenv` this new virtualenv within Tox's virtualenv does **not** use the Python interpreter that Tox provides but the system's interpreter (or the one wherever you run tox which is a conda environment in my case). 

If these statements hold we could fix the problem by telling `pytest-virtualenv` somehow that it should use the Python interpreter provided in the environment created by Tox. Correct?

Judging from the implementation of `pytest-virtualenv` there is an official [config hook](https://github.com/manahl/pytest-plugins/blob/master/pytest-virtualenv/pytest_virtualenv.py#L27) which seems to provide exactly this functionality. Right now it defaults to using the system's Python interpreter. Couldn't we somehow use this to solve this issue? I am no expert in Tox but surely there is a way to determine the Python interpreter of Tox while `conftest.py` is read. This would allow us to modify the `venv` fixture so that it uses the `virtualenv` fixture of `pytest-virtualenv` but with `virtualenv_executable` set to Tox's Python interpreter. Does this sound reasonable to you?

As you see I would rather prefer your suggested solution (1) if it is practical since I don't want to lose the ability to get the coverage of these unit tests. I agree that your PR would give as much more control than relying on `pytest-virtualenv` but with more control and thus great power comes great responsibility. What would Peter Parker do? ;-)

**Comment by FlorianWilhelm** [comment]
@abravalheri, I hope to be able to work on this and look into more details on Tuesday.

**Comment by FlorianWilhelm** [comment]
This is fixed in version 3.2.3.
